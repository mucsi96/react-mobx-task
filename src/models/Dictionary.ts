import Transformation from './Transformation';
import { observable, action } from 'mobx';

export default class Dictionary {
  public readonly id: string;

  @observable
  public name: string;

  @observable
  public transformations: Transformation[] = [];

  constructor(name: string) {
    this.id = Date.now().toString();
    this.name = name;
  }

  @action
  public addTransformation(transformation: Transformation) {
    this.transformations.push(transformation);
  }

  @action
  public removeTransformation(transformationId: string) {
    const index = this.transformations.findIndex(
      ({ id }) => id === transformationId
    );

    if (index < 0) {
      throw new Error(
        `Transformation with id "${transformationId}" was not found.`
      );
    }

    this.transformations.splice(index, 1);
  }

  public getTransformationById(transformationId: string): Transformation {
    const transformation = this.transformations.find(
      ({ id }) => id === transformationId
    );

    if (!transformation) {
      throw new Error(
        `Transformation with id ${transformationId} was not found.`
      );
    }

    return transformation;
  }
}
