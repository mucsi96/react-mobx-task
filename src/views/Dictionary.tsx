import * as React from 'react';
import { View } from '.';

export default class Dictionary implements View {
  public path = '/dictionary';

  public static pattern = '/dictionary';

  public render() {
    return <span>{'Dictionary'}</span>;
  }

  public createFromParams() {
    return new Dictionary();
  }
}
