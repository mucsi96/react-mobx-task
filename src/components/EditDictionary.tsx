import * as React from 'react';
import { inject, observer } from 'mobx-react';
import DictionaryStore from '../stores/DictionaryStore';
import ViewStore from '../stores/ViewStore';
import { observable, action } from 'mobx';
import Dictionary from '../models/Dictionary';

interface EditDictionaryProps {
  dictionaryId?: string;
}

@inject('dictionaryStore', 'viewStore')
@observer
export default class EditDictionary extends React.Component<
  EditDictionaryProps
> {
  get injected() {
    return (this.props as unknown) as {
      dictionaryStore: DictionaryStore;
      viewStore: ViewStore;
    };
  }

  @observable
  private name: string = '';

  private readonly dictionary?: Dictionary;

  constructor(props: EditDictionaryProps) {
    super(props);
    if (props.dictionaryId) {
      this.dictionary = this.injected.dictionaryStore.getDictionaryById(
        props.dictionaryId
      );
      this.name = this.dictionary.name;
    }
  }

  public render(): JSX.Element {
    return (
      <form className="ui form" onSubmit={this.handleSubmit}>
        <h4 className="ui header">
          {this.dictionary ? 'Edit dictionary' : 'Create dictionary'}
        </h4>
        <div className="field">
          <label>Name</label>
          <input
            autoFocus={true}
            type="text"
            value={this.name}
            onChange={event => this.setName(event.target.value)}
          />
        </div>
        <button type="submit" className="ui button primary">
          {this.dictionary ? 'Save' : 'Create'}
        </button>
        <button type="button" className="ui button" onClick={this.handleCancel}>
          Cancel
        </button>
      </form>
    );
  }

  private handleSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    if (!this.dictionary) {
      this.injected.dictionaryStore.addDictionary(new Dictionary(this.name));
      this.injected.viewStore.showOverview();
    } else {
      this.dictionary.name = this.name;
      this.injected.viewStore.showOverview();
    }
  };

  private handleCancel = () => {
    this.injected.viewStore.showOverview();
  };

  @action
  private setName(name: string) {
    this.name = name;
  }
}
