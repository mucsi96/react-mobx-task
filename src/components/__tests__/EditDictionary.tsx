import * as React from 'react';
import * as RTL from 'react-testing-library';
import { Provider } from 'mobx-react';
import EditDictionary from '../EditDictionary';
import Dictionary from '../../models/Dictionary';

const testDictionary = {
  id: '1',
  name: 'a',
  transformations: [
    { id: '1', from: 'b', to: 'c' },
    { id: '2', from: 'd', to: 'e' }
  ]
};

const dictionaryStore = {
  addDictionary: jest.fn(),
  getDictionaryById: (id: string) => (id === '7' ? testDictionary : undefined)
};

const viewStore = {
  showOverview: jest.fn()
};

const render = ({ create = false } = {}) =>
  RTL.render(
    <Provider dictionaryStore={dictionaryStore} viewStore={viewStore}>
      <EditDictionary dictionaryId={create ? undefined : '7'} />
    </Provider>
  );

describe('EditDictionary', () => {
  it('renders', () => {
    const { asFragment } = render();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders in creation case', () => {
    const { asFragment } = render({ create: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('creates new dictionary on submit in case of creation', () => {
    dictionaryStore.addDictionary.mockClear();
    const { getByLabelText, getByText } = render({ create: true });
    RTL.fireEvent.change(getByLabelText('Name'), {
      target: { value: 'new dict' }
    });
    RTL.fireEvent.submit(getByText('Create'));
    expect(dictionaryStore.addDictionary).toBeCalled();
  });

  it('modifies dictionary on submit in case of editing', () => {
    const { getByLabelText, getByText } = render();
    RTL.fireEvent.change(getByLabelText('Name'), {
      target: { value: 'new name' }
    });
    RTL.fireEvent.submit(getByText('Save'));
    expect(testDictionary.name).toEqual('new name');
  });
  it('navigates to overview on "Cancel" click', () => {
    viewStore.showOverview.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Cancel'));
    expect(viewStore.showOverview).toBeCalled();
  });
});
