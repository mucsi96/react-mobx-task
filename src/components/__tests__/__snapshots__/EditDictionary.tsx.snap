// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDictionary renders 1`] = `
<DocumentFragment>
  <div
    class="ui basic segment"
  >
    <form
      class="ui form"
    >
      <div
        class="ui header"
      >
        Edit dictionary
      </div>
      <div
        class="field"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          id="name"
          type="text"
          value="a"
        />
      </div>
      <button
        class="ui button primary"
        type="submit"
      >
        Save
      </button>
      <button
        class="ui button"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`EditDictionary renders in creation case 1`] = `
<DocumentFragment>
  <div
    class="ui basic segment"
  >
    <form
      class="ui form"
    >
      <div
        class="ui header"
      >
        Create dictionary
      </div>
      <div
        class="field"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          id="name"
          type="text"
          value=""
        />
      </div>
      <button
        class="ui button primary"
        type="submit"
      >
        Create
      </button>
      <button
        class="ui button"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
